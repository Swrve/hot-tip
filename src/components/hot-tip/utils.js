const getBottom = (bounds, force) => {
  const {left, top, width, height, clientWidth, clientHeight} = bounds

  if (!force && top + height + 40 >= clientHeight) {
    return getTop(bounds, true) // eslint-disable-line
  }
  const x = left + width / 2,
    y = top + height,
    leftOffset = x < 100,
    rightOffset = !leftOffset && x + 100 > clientWidth,
    layoutClass = leftOffset ? 'left-offset' : (rightOffset ? 'right-offset' : 'centred') // eslint-disable-line

  return {
    x,
    y,
    positionClass: 'bottom',
    layoutClass
  }
}

const getTop = (bounds, force) => {
  const {top} = bounds,
    bottom = getBottom(bounds, true)

  if (!force && top - 40 <= 0) {
    return bottom
  }

  return Object.assign({}, bottom, {
    y: top,
    positionClass: 'top'
  })
}

const getRight = (bounds, force) => {
  const {left, top, width, height, clientWidth, clientHeight} = bounds

  if (!force && left + width + 200 >= clientWidth) {
    return getLeft(bounds, true) // eslint-disable-line
  }
  const x = left + width,
    y = top + height / 2,
    topOffset = y < 50,
    bottomOffset = !topOffset && y + 50 > clientHeight,
    layoutClass = topOffset ? 'top-offset' : (bottomOffset ? 'bottom-offset' : 'mid') // eslint-disable-line

  return {
    x,
    y,
    positionClass: 'right',
    layoutClass
  }
}

const getLeft = (bounds, force) => {
  const {left} = bounds,
    right = getRight(bounds, force)

  if (!force && left - 200 <= 0) {
    return right
  }

  return Object.assign({}, right, {
    x: left,
    positionClass: 'left'
  })
}

export const getPosition = (position) => {
  switch (position) {
  case ('top'):
    return getTop
  case ('right'):
    return getRight
  case ('left'):
    return getLeft
  case ('bottom'):
  default:
    return getBottom
  }
}

export const getBounds = (e) => {
  const {top, left, width, height} = e.currentTarget.getBoundingClientRect(),
    {pageXOffset, pageYOffset} = window,
    {scrollTop, clientTop, scrollLeft, clientLeft, clientWidth, clientHeight} = document.body,
    absTop = top + Math.max(pageYOffset, scrollTop) - clientTop,
    absLeft = left + Math.max(pageXOffset, scrollLeft) - clientLeft

  return {
    top: absTop,
    left: absLeft,
    width,
    height,
    clientWidth,
    clientHeight
  }
}
