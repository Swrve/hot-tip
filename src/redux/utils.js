import {combineReducers} from 'redux'
import hotTipReducer from './reducer'

export default function addHotTipToStore(store) {
  return combineReducers(
    hotTipReducer,
    ...store
  )
}
