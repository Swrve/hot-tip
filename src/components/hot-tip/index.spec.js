import React from 'react'
import HotTip from './'
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store'
import * as utils from './utils';

const middlewares = []
const mockStore = configureStore(middlewares)

describe('Components', function() {
  describe('HotTip', function() {
    let store
    beforeEach(() => {
      store = mockStore({})
    })

    it('renders', () => {
      const component = renderer.create(
          <HotTip store={store} text="click me"><a>{'link'}</a></HotTip>
        ),
        tip = component.toJSON()

      expect(tip.type).toEqual('a')
      expect(tip.children[0]).toEqual('link')
      expect(tip).toMatchSnapshot()
    })

    it('hover', () => {
      utils.getBounds = jest.fn((one) => ({one}))
      utils.getPosition = jest.fn((one) => (two) => ({one, two}))
      const component = renderer.create(
        <HotTip store={store} text="click me"><a>{'link'}</a></HotTip>
      )
      let tip = component.toJSON()

      // manually trigger the callback
      tip.props.onMouseOver('hello');
      expect(utils.getBounds).toBeCalledWith('hello')
      expect(utils.getPosition).toBeCalledWith('bottom')

      tip = component.toJSON();
      expect(tip).toMatchSnapshot();

      tip.props.onMouseOut();

      tip = component.toJSON();
      expect(tip).toMatchSnapshot();

      const actions = store.getActions()

      expect(actions[0].payload.position).toEqual({
        one: 'bottom',
        two: {one: 'hello'}
      })
      expect(actions[1].payload).toEqual({visible: false})
    })
  })
})
