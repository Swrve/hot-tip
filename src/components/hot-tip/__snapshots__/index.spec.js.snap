// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components HotTip hover 1`] = `
<a
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  link
</a>
`;

exports[`Components HotTip hover 2`] = `
<a
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  link
</a>
`;

exports[`Components HotTip renders 1`] = `
<a
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  link
</a>
`;
